dali/operators/reader/loader/nemo_asr_loader.cc

```diff
@@ -0,0 +1,212 @@
+  }
+
+  DALI_ENFORCE(Size() > 0, "No files found.");
+  if (shuffle_) {
+    // seeded with hardcoded value to get
+    // the same sequence on every shard
+    std::mt19937 g(kDaliDataloaderSeed);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+  Reset(true);
+}
+
+void NemoAsrLoader::Reset(bool wrap_to_shard) {
+  current_index_ = wrap_to_shard ? start_index(shard_id_, num_shards_, Size()) : 0;
+  current_epoch_++;
+
+  if (shuffle_after_epoch_) {
+    std::mt19937 g(kDaliDataloaderSeed + current_epoch_);
+    std::shuffle(entries_.begin(), entries_.end(), g);
+  }
+}
+
+void NemoAsrLoader::PrepareEmpty(AsrSample &sample) {
+  PrepareEmptyTensor(sample.audio_);
+}
+
+template <typename OutputType, typename DecoderOutputType>
+void NemoAsrLoader::ReadAudio(Tensor<CPUBackend> &audio,
+                              const AudioMetadata &audio_meta,
+                              const NemoAsrEntry &entry,
+                              AudioDecoderBase &decoder,
+                              std::vector<uint8_t> &decode_scratch,
+                              std::vector<float> &resample_scratch) {
+  audio.set_type(TypeTable::GetTypeInfo(dtype_));
+  auto shape = DecodedAudioShape(audio_meta, sample_rate_, downmix_);
+  assert(shape.size() > 0);
+  audio.Resize(shape);
+
+  bool should_resample = sample_rate_ > 0 && audio_meta.sample_rate != sample_rate_;
+  bool should_downmix = audio_meta.channels > 1 && downmix_;
+
+  int64_t decode_scratch_sz = 0;
+  int64_t resample_scratch_sz = 0;
+  if (should_resample || should_downmix || dtype_ != DALI_INT16)
+    decode_scratch_sz = audio_meta.length * audio_meta.channels;
+
+  decode_scratch.resize(decode_scratch_sz * sizeof(DecoderOutputType));
+
+  // resample scratch is used to prepare a single or multiple (depending if
+  // downmixing is needed) channel float input, required by the resampling
+  // kernel
+  int64_t out_channels = should_downmix ? 1 : audio_meta.channels;
+  if (should_resample)
+    resample_scratch_sz = audio_meta.length * out_channels;
+
+  resample_scratch.resize(resample_scratch_sz);
+
+  DecodeAudio<OutputType, DecoderOutputType>(
+    view<OutputType>(audio), decoder, audio_meta, resampler_,
+    {reinterpret_cast<DecoderOutputType *>(decode_scratch.data()), decode_scratch_sz},
+    {resample_scratch.data(), resample_scratch_sz},
+    sample_rate_, downmix_,
+    entry.audio_filepath.c_str());
+}
+
+void NemoAsrLoader::ReadSample(AsrSample& sample) {
+  auto &entry = entries_[current_index_++];
+
+  // handle wrap-around
+  MoveToNextShard(current_index_);
+
+  // metadata info
+  sample = {};
+  DALIMeta meta;
+  meta.SetSourceInfo(entry.audio_filepath);
+  meta.SetSkipSample(false);
+  sample.audio_.SetMeta(meta);
+  sample.text_ = entry.text;
+
+  // Ignoring copy_read_data_, Sharing data is not supported with this loader
+
+  bool use_resampling = sample_rate_ > 0;
+  DALIDataType decode_type = use_resampling ? DALI_FLOAT : dtype_;
```

Comment:
We'll read to float and possibly convert back even if resampling is not actually required :(

Reply:
Yes, I know. It's a limitation with the current design of the decoder (Needs to know the type at construction and not at decode time)